src/profile-logic/processed-profile-versioning.js

```diff
@@ -468,28 +471,59 @@ const _upgraders = {
             },
             time: data.endTime,
             name: markers.name[i],
-          };
-        } else {
-          newDataArray[i] = data;
+          });
         }
       }
-      thread.markers.data = newDataArray;
 
-      // Adding the `end` markers that we created to appropriate indices.
-      // Markers need to be sorted by time.
-      for (const i in markersToAdd) {
-        let index = parseInt(i);
-        while (
-          thread.markers.time.length > index &&
-          markersToAdd[i].time >= thread.markers.time[index]
-        ) {
-          index++;
-        }
+      extraMarkers.sort((a, b) => a.time - b.time);
 
-        thread.markers.data.splice(index, 0, markersToAdd[i].data);
-        thread.markers.time.splice(index, 0, markersToAdd[i].time);
-        thread.markers.name.splice(index, 0, markersToAdd[i].name);
+      // Create a new markers table that includes both the old markers and the
+      // markers from extraMarkers, sorted by time.
+      const newMarkers = {
+        length: 0,
+        name: [],
+        time: [],
+        data: [],
+      };
+
+      // We compute the new markers list by doing one forward pass. Both the
+      // old markers (stored in |markers|) and the extra markers are already
+      // sorted by time.
+
+      let nextOldMarkerIndex = 0;
+      let nextOldMarkerTime = markers.time[0];
+      let nextExtraMarkerIndex = 0;
+      let nextExtraMarkerTime =
+        extraMarkers.length > 0 ? extraMarkers[0].time : Infinity;
```

Comment:
If `extraMarkers.length === 0` couldn't we just use the old table and skip iterating over all the markers again, or am I missing something?

Reply:
Good point! I'll make that change.